Refactor Admin Events Index

juliannatetreault opened the pull request:
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
**The issue**:
> As a Forem admin setting up events, it's pretty unwieldy to open /admin/events because everything is on one page. That includes past events, future events, the event creation page, etc. 

**The solution**:
This PR refactors the `/admin/events` index page by removing the event creation and event edit forms from the index into their own views and by cleaning up the `/admin/events` UI by separating out "Upcoming Events" and "Past Events" by use of a crayons table. Additionally, this PR adds a "New Event" button to the top of the `/admin/events` index and inline "Edit" buttons next to each event for easy editing. This PR also refactors the `Admin::EventsController`, Admin event-related forms, and Admin event-related specs. It also adds additional actions to the `Admin::EventsController` and additional routes to `routes.rb` (`#edit`, `#new`, and `#update`) and adds and refactors tests around creating and updating events.

## Related Tickets & Documents
Closes https://github.com/forem/InternalProjectPlanning/issues/208

## QA Instructions, Screenshots, Recordings
**To QA this PR, first ensure that you have an `admin` role, and then please do the following**:

- Navigate to `/admin/events` to see the new, Pages-inspired index page:
![Screen Shot 2020-10-26 at 2 28 51 PM](https://user-images.githubusercontent.com/32834804/97224981-93acad80-1797-11eb-95c8-95ed6bf1690d.png)

- Click on "New Event" to create a new event. After filling out the form, click "Create Event" to create your new event. **Note**: an image should only appear in the `/admin/events` index if you upload one, otherwise, no image should display in the table (see final QA screenshot to see what that looks like):
![Screen Shot 2020-10-26 at 2 31 08 PM](https://user-images.githubusercontent.com/32834804/97225161-e5553800-1797-11eb-9082-e1b5fb4d1f05.png)

- You should be redirected back to `/admin/events` upon the successful creation of an event. You should see a flash message indicating that the creation was successful and you should also see your event under the correct heading, "Upcoming Events" or "Past Events," depending on when your event takes place:
![Screen Shot 2020-10-26 at 2 34 04 PM](https://user-images.githubusercontent.com/32834804/97225466-4da41980-1798-11eb-8c5a-6b10a3b51cdc.png)

- Click the "Edit" button next to your newly created event to edit the event and edit one or all of the fields. Upon a successful update, you should be redirected back to `/admin/events` and you should see a flash message indicating that the update was successful:
![Screen Shot 2020-10-26 at 2 35 32 PM](https://user-images.githubusercontent.com/32834804/97225640-8217d580-1798-11eb-825e-944e5fd4fdc9.png)

- After creating and updating multiple events, your `/admin/events` index should look something like this, with the events properly sorted under the correct headers, "Upcoming Events" and "Past Events":
![Screen Shot 2020-10-26 at 2 25 10 PM](https://user-images.githubusercontent.com/32834804/97224649-108b5780-1797-11eb-8e2a-9fc75ffe5d74.png)

- Any published events can be seen on `/events`:
![Screen Shot 2020-10-26 at 2 49 24 PM](https://user-images.githubusercontent.com/32834804/97227271-dcb23100-179a-11eb-9434-ac7a6e9526af.png)

- Try to break things! ⚒️ 

## Added tests?

- [x] Yes, and updated existing specs
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Added to documentation?

- [ ] Docs.forem.com
- [ ] README
- [x] No documentation needed

## [optional] Are there any post-deployment tasks we need to perform?
None that I can think of!

## [optional] What gif best describes this PR or how it makes you feel?

![The word "event" in various fonts](https://media.giphy.com/media/ekY8JdlX7FqQTbteE7/giphy.gif)

juliannatetreault commented on app/views/admin/pages/index.html.erb:
```diff
@@ -31,7 +31,7 @@
           <div class="flex py-2 items-center">
               <%= link_to "Privacy Policy", privacy_path %>
-              <%= link_to "Override", new_admin_page_path(slug: "privacy"), class: "ml-auto crayons-btn crayons-btn--s crayons-btn--danger"  %>
+              <%= link_to "Override", new_admin_page_path(slug: "privacy"), class: "ml-auto crayons-btn crayons-btn--s crayons-btn--danger" %>
```
This just removes an extra space ✂️

juliannatetreault commented on app/views/admin/pages/index.html.erb:
```diff
@@ -16,7 +16,7 @@
 <div class="mt-8 pages__override_defaults">
   <div class="my-6">
-    <% if !@code_of_conduct || !@privacy || !@terms%>
+    <% if !@code_of_conduct || !@privacy || !@terms %>
```
This just adds a space after `!@terms`.

juliannatetreault commented on app/controllers/admin/events_controller.rb:
```diff
@@ -4,35 +4,40 @@ class EventsController < ApplicationController
     include ApplicationHelper
 
     def index
+      @events = Event.all.order(starts_at: :desc).page(params[:page]).per(20)
```
How do we feel about this pagination? Is it too much or too little per page?

mstruve replied:
Luckily it looks like `.all` is ignored here with the pagination but we should probably remove it anyways since in normal circumstances `.all` would pull everything out of the database. Surprised it doesnt in this case, I found this interesting 👇 
```ruby
[55] pry(main)> Event.all.order(starts_at: :desc).page(0).per(1)
  Event Load (8.1ms)  SELECT "events".* FROM "events" ORDER BY "events"."starts_at" DESC LIMIT $1 OFFSET $2  [["LIMIT", 1], ["OFFSET", 0]]
=> []
[56] pry(main)> Event.order(starts_at: :desc).page(0).per(1)
  Event Load (1.7ms)  SELECT "events".* FROM "events" ORDER BY "events"."starts_at" DESC LIMIT $1 OFFSET $2  [["LIMIT", 1], ["OFFSET", 0]]
=> []
```

As for page size, 20 seems good to me.

mstruve commented on app/controllers/admin/events_controller.rb:
```diff
@@ -4,35 +4,40 @@ class EventsController < ApplicationController
       else
         flash[:danger] = @event.errors.full_messages
-        render "index.html.erb"
+        render new_admin_event_path
```
Nice cleanup!

mstruve reviewed (approved):
Works great! Nothing blocking from me 🚀 

Would be nice to put a light line between 2 events maybe so they were easier to follow across the page. Just a thought 
![Screen Shot 2020-10-26 at 5 39 05 PM](https://user-images.githubusercontent.com/1813380/97231690-dc6d6200-17a9-11eb-9f77-29efd9447c0f.png)

benhalpern reviewed (approved):
Looks good to me. Nice improvements.

juliannatetreault commented:
@mstruve I removed the `.all` from `#index` in the `Admin:::EventsController` and added an `<hr>` tag between each event on the `/admin/events` table to create some extra separation between events. 😊

ludwiczakpawel commented on app/views/admin/events/_event_form.html.erb:
```diff
@@ -1,50 +1,54 @@
+      <%= f.file_field :profile_image, class: "form-control" %>
+      <img src="<%= @event.profile_image_url %>" style="width: 25%;" alt="event profile image">
+    </div>
+    <div class="form-group">
+      <%= f.label :title %>
+      <%= f.text_field :title, maxlength: 90, size: 40, required: true, class: "form-control" %>
+    </div>
```
For future (well, you can also do it here if you want to) we could reuse Crayons markup for form fields... So this one in particular would look as follows:

```suggestion
<div class="crayons-field">
  <%= f.label :title, class: "crayons-field__label" %>
  <%= f.text_field :title, maxlength: 90, size: 40, required: true, class: "crayons-textfield" %>
</div>
```

^ and that applies to all form fields.

ludwiczakpawel commented on app/views/admin/events/_event_form.html.erb:
```diff
@@ -1,50 +1,54 @@
+      <%= f.label :live_now %>
+      <%= f.check_box :live_now %>
+    </div>
+    <%= f.submit class: "btn btn-primary" %>
```
This could become crayons button as well:

```suggestion
    <%= f.submit class: "crayons-btn" %>
```

juliannatetreault replied:
Also, done!

juliannatetreault commented:
@ludwiczakpawel I've made your suggested changes! Thanks so much for the review. 😄

Zhao-Andy reviewed (approved):
LGTM!
